### PR TITLE
Makes a global list of all items

### PR DESCRIPTION
**Pull request by coiax**
Iterating through `world` is bad, and I think other projects/plans would
like to operate over all items as well. So let's have a global list,
trading memory for performance.

- Also, shuffles the list when performing RPG loot, because otherwise I
have a niggling suspicion that the item fortification scrolls may be
placed in a slightly more predictable pattern then they should.

### CODE REVIEW

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -116,6 +116,8 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 	if(GLOB.rpg_loot_items)
 		rpg_loot = new(src)
 
+	GLOB.all_items_list[src] = TRUE
```
Comment (Cyberboss) [inline code/game/objects/items.dm]: += is faster I believe. @RemieRichards?

**Comment by xxalpha** [comment]
When do you ever need to access a global list of ALL items?

**Comment by Iamgoofball** [comment]
Beats doing world

**Comment by xxalpha** [comment]
It really doesn't beat that, the only thing that it is useful for (arguably) is an event that might not happen for many rounds, but this list will always exist and always bloat the memory usage. This is not a good trade off, just let the CPU do the occasional work rather than always have a huge list that will only be occasionally used.

**Comment by coiax** [comment]
I know that @kevinz000 directional gravity also deals with a list of all items. In addition, I have some potential features (floating for items in no gravity, hyperspace turbulance) that would benefit from lists of all items. Lemmie do some testing to determine the size of the list.

**Comment by coiax** [comment]
On testing on Box, the `all_items_list` has 29,220 items.

**Comment by optimumtact** [comment]
Do you have any proof this is faster than doing a typed for? we know the byond world list is special and has optimisations which might include shortcuts for common types like this

**Comment by RemieRichards** [comment]
@optimumtact ```/obj/item``` is not a built in type, it has no optimisations.

File: code/game/objects/items.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: that's correct.
```
L += V
L[V] = *
L |= V
```
is the order of fastest->slowest

Comment (coiax) [inline code/game/objects/items.dm]: so wait, do I go `GLOB.all_items_list += TRUE` then?

**Comment by optimumtact** [comment]
yeah you're probably right

File: code/game/objects/items.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/objects/items.dm]: no lol.
look at ```V``` in my example, it's the thing being added.

you want to add ```src``` to the list, what on earth is adding total_number_of_item ```TRUE```s going to do for you?

**Comment by duncathan** [comment]
we have lists of similar magnitude to this so it *should* be fine-ish

**Comment by AnturK** [comment]
Not sure about this, feels like catch-all without real purpose right now.

File: code/game/objects/items.dm (hunk 1 above)
Comment (coiax) [inline code/game/objects/items.dm]: derp

**Comment by kevinz000** [comment]
@coiax gravity is more optimized and doesn't check every item ever anymore, because that method made it horribly demanding and laggy.

**Comment by coiax** [comment]
I wanted to make all items on the station zlevel float when gravity cuts out.

**Comment by kevinz000** [comment]
so do a world loop there's already talk about memory being too scarce why would you do this?

**Comment by KorPhaeron** [comment]
@tgstation/commit-access 

None of you seemed very interested and its been a week so I am closing it, feel free to reopen

**Pull request closed without merging**